src/Microsoft.AspNetCore.Identity/UserClaimsPrincipalFactoryOfTUser.cs

```diff
@@ -0,0 +1,94 @@
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="UserClaimsPrincipalFactory{TUser}"/> class.
+        /// </summary>
+        /// <param name="userManager">The <see cref="UserManager{TUser}"/> to retrieve user information from.</param>
+        /// <param name="optionsAccessor">The configured <see cref="IdentityOptions"/>.</param>
+        public UserClaimsPrincipalFactory(
+            UserManager<TUser> userManager, 
+            IOptions<IdentityOptions> optionsAccessor)
+        {
+            if (userManager == null)
+            {
+                throw new ArgumentNullException(nameof(userManager));
+            }
+            if (optionsAccessor == null || optionsAccessor.Value == null)
+            {
+                throw new ArgumentNullException(nameof(optionsAccessor));
+            }
+            UserManager = userManager;
+            Options = optionsAccessor.Value;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="UserManager{TUser}"/> for this factory.
+        /// </summary>
+        /// <value>
+        /// The current <see cref="UserManager{TUser}"/> for this factory instance.
+        /// </value>
+        public UserManager<TUser> UserManager { get; private set; }
+
+        /// <summary>
+        /// Gets the <see cref="IdentityOptions"/> for this factory.
+        /// </summary>
+        /// <value>
+        /// The current <see cref="IdentityOptions"/> for this factory instance.
+        /// </value>
+        public IdentityOptions Options { get; private set; }
+
+        /// <summary>
+        /// Creates a <see cref="ClaimsPrincipal"/> from an user asynchronously.
+        /// </summary>
+        /// <param name="user">The user to create a <see cref="ClaimsPrincipal"/> from.</param>
+        /// <returns>The <see cref="Task"/> that represents the asynchronous creation operation, containing the created <see cref="ClaimsPrincipal"/>.</returns>
+        public virtual async Task<ClaimsPrincipal> CreateAsync(TUser user)
+        {
+            if (user == null)
+            {
+                throw new ArgumentNullException(nameof(user));
+            }
+            var userId = await UserManager.GetUserIdAsync(user);
+            var userName = await UserManager.GetUserNameAsync(user);
+            var id = new ClaimsIdentity(Options.Cookies.ApplicationCookieAuthenticationScheme,
+                Options.ClaimsIdentity.UserNameClaimType,
+                Options.ClaimsIdentity.RoleClaimType);
+            id.AddClaim(new Claim(Options.ClaimsIdentity.UserIdClaimType, userId));
+            id.AddClaim(new Claim(Options.ClaimsIdentity.UserNameClaimType, userName));
+            if (UserManager.SupportsUserSecurityStamp)
+            {
+                id.AddClaim(new Claim(Options.ClaimsIdentity.SecurityStampClaimType, 
+                    await UserManager.GetSecurityStampAsync(user)));
+            }
+            if (UserManager.SupportsUserRole)
```

Comment:
I am not sure about this. Is it still valuable to check this and add roles into the claims?